code/parse/sexp.cpp

```diff
@@ -2681,159 +2701,120 @@ int check_sexp_syntax(int node, int return_type, int recursive, int *bad_node, i
 				break;
 
 			case OPF_GOAL_NAME:
+			case OPF_EVENT_NAME:
 				if (type2 != SEXP_ATOM_STRING)
 					return SEXP_CHECK_TYPE_MISMATCH;
 
 				// we only need to check the campaign list if running in Fred and are in campaign mode.
-				// otherwise, check the set of current goals
-				if ( Fred_running && (mode == SEXP_MODE_CAMPAIGN) ) {
+				// otherwise, check the set of current goals/events
+				if (Fred_running && (mode == SEXP_MODE_CAMPAIGN)) {
 					z = find_parent_operator(node);
 					Assert(z >= 0);
 					z = Sexp_nodes[z].rest;  // first argument of operator should be mission name
 					Assert(z >= 0);
-					for (i=0; i<Campaign.num_missions; i++)
+
+					// can't check special-arg
+					if (Sexp_nodes[z].flags & SNF_SPECIAL_ARG_IN_NODE)
+						break;
+
+					// look for mission
+					for (i = 0; i < Campaign.num_missions; i++)
 						if (!stricmp(CTEXT(z), Campaign.missions[i].name))
 							break;
 
+					if (i >= Campaign.num_missions) {
+						if (bad_node)
+							*bad_node = z;
+
+						return SEXP_CHECK_INVALID_MISSION_NAME;
+					}
+
 					// read the goal/event list from the mission file if both num_goals and num_events
 					// are < 0
-					if ((Campaign.missions[i].num_goals <= 0) && (Campaign.missions[i].num_events <= 0) )
+					if ((Campaign.missions[i].num_goals <= 0) && (Campaign.missions[i].num_events <= 0))
 						read_mission_goal_list(i);
-					
-					if (i < Campaign.num_missions) {
-						for (t=0; t<Campaign.missions[i].num_goals; t++)
+
+					if (type == OPF_GOAL_NAME) {
+						for (t = 0; t < Campaign.missions[i].num_goals; t++)
 							if (!stricmp(CTEXT(node), Campaign.missions[i].goals[t].name))
 								break;
 
 						if (t == Campaign.missions[i].num_goals)
 							return SEXP_CHECK_INVALID_GOAL_NAME;
```

Comment:
What about setting `*bad_node` here?

Is that not needed if FRED is running? Even if so, maybe we should defensively set it anyway (if`bad_node` is non-null of course), in case the logic changes in the future such that FSO makes these checks as well?

Reply:
Setting `bad_node` is only necessary if the bad node is something other than the node being checked.  Elsewhere in the function it assumes the bad node is the current node.